File: library/src/main/java/de/nenick/espressomacchiato/elements/EspView.java
package de.nenick.espressomacchiato.elements;

import android.support.test.espresso.ViewInteraction;
import android.support.test.espresso.action.ViewActions;
import android.view.View;

import org.hamcrest.Matcher;

import java.util.ArrayList;
import java.util.Arrays;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.isEnabled;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static org.hamcrest.CoreMatchers.allOf;
import static org.hamcrest.CoreMatchers.not;

public class EspView {

    protected final Matcher<View> baseMatcher;

    public static EspView byId(int resourceId) {
        return new EspView(resourceId);
    }

    public EspView(int resourceId) {
        this.baseMatcher = withId(resourceId);
    }

    public EspView(Matcher<View> baseMatcher) {
        this.baseMatcher = baseMatcher;
    }

    public void assertIsVisible() {
        findView().check(matches(isDisplayed()));
    }

    public void assertIsHidden() {
        findView().check(matches(not(isDisplayed())));
    }

    public void assertIsEnabled() {
        findView().check(matches(isDisplayed()));
    }

    public void assertIsDisabled() {
        findView().check(matches(not(isEnabled())));
    }

    public void click() {
        findView(isDisplayed()).perform(ViewActions.click());
    }

    @SafeVarargs
    protected final ViewInteraction findView(Matcher<View>... additionalMatcher) {
        ArrayList<Matcher<? super View>> allMatcher = new ArrayList<>();
        allMatcher.add(baseMatcher);
        allMatcher.addAll(Arrays.asList(additionalMatcher));
        return onView(allOf(allMatcher));
    }
}

File: esp-android/src/main/java/de/nenick/espressomacchiato/view/EspView.kt
package de.nenick.espressomacchiato.view

import android.view.View
import androidx.annotation.IdRes
import androidx.test.espresso.Espresso
import androidx.test.espresso.Root
import androidx.test.espresso.ViewAction
import androidx.test.espresso.ViewInteraction
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.RootMatchers
import androidx.test.espresso.matcher.ViewMatchers.withId
import de.nenick.espressomacchiato.EspGlobalSettings
import de.nenick.espressomacchiato.internals.*
import de.nenick.espressomacchiato.internals.IsActivityInForegroundMatcher.Companion.assertActivityInForeground
import org.hamcrest.Matcher

// Scope marker reduces the amount methods when views get nested e.g. the page style.
// Result is in sub lambda blocks you don't have access to the parent lambda block methods.
@ScopeDslMarker
@OpenForExtensions
class EspView(

        /** The logic to locate this view. */
        val viewMatcher: Matcher<View>,

        /**
         * Optional to located views inside different roots than the activity.
         *
         * E.g. dialogs, see [RootMatchers] for more details.
         */
        private val rootMatchers: Matcher<Root>? = null,

        /** Short way to call assertions and actions on this view. */
        interactions: EspView.() -> Unit = {}

) : HasChecks, HasActions, HasInteractions {

    /** Short way to locate this view by his id, the most common way to do that. */
    constructor(@IdRes viewId: Int, interactions: EspView.() -> Unit = {})
            : this(withId(viewId), null, interactions)

    /** Short way to locate this view by his id within different root (e.g. dialogs). */
    constructor(@IdRes viewId: Int, rootMatchers: Matcher<Root>, interactions: EspView.() -> Unit = {})
            : this(withId(viewId), rootMatchers, interactions)

    /** Short way to locate this view by custom view matcher. */
    constructor(viewMatcher: Matcher<View>, interactions: EspView.() -> Unit = {})
            : this(viewMatcher, null, interactions)

    init {
        // The compile may be right to complain about it but the possible issues did not happen yet.
        // This supports the fancy EspView(id) { check(something) } style.
        @Suppress("LeakingThis")
        interactions(this)
    }

    override fun onView(): ViewInteraction {
        val viewInteraction = Espresso.onView(viewMatcher)

        if (rootMatchers != null) {
            viewInteraction.inRoot(rootMatchers)
        } else if (EspGlobalSettings.activityForegroundCheckEnabled) {
            assertActivityInForeground("Found something different than an activity in foreground. Is this a wrong situation or did you forgot to add a root matcher e.g. isDialog?")
        }

        return viewInteraction
    }

    override fun check(matcher: Matcher<View>) {
        onView().check(ViewAssertions.matches(matcher))
    }

    override fun perform(action: ViewAction) {
        onView().perform(action)
    }

}
